### PR TITLE
Invoice details listing page is created

### PR DESCRIPTION
**Pull request by Ayush8923**
Targets #1324 #1325

### CODE REVIEW

File: Modules/Invoice/Resources/views/invoice-details-listing.blade.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
+@extends('invoice::layouts.master')
+@section('content')
+
+<div class="mx-15" id="vueContainer">
+    <br>
+    <br>
+
+    <div class="d-flex justify-content-between mb-2">
+        <h4 class="mb-1 pb-1">              Invoice Details</h4>
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/invoice-details-listing.blade.php:9]:
```suggestion
        <h4 class="mb-1 pb-1">Invoice Details</h4>
```

File: Modules/Invoice/Routes/web.php [hunk 2]
```diff
@@ -23,4 +23,5 @@
     Route::post('/{invoiceId}/update', 'InvoiceController@update')->name('invoice.update');
     Route::delete('/{invoiceId}/delete', 'InvoiceController@destroy')->name('invoice.delete');
     Route::post('/{invoice}', 'InvoiceController@sendEmail')->name('invoice.sendEmail');
+    Route::get('/invoice-details-listing', 'InvoiceController@invoiceDetails')->name('invoice.invoice-details');
```
Comment (GauravGusain98) [inline Modules/Invoice/Routes/web.php:26]:
```suggestion
    Route::get('/invoice-details-listing', 'InvoiceController@invoiceDetails')->name('invoice.details');
```

File: resources/views/layouts/navbar.blade.php [hunk 3]
```diff
@@ -62,6 +62,7 @@
                     <a class="dropdown-item" href="{{ route('invoice.index') }}">Invoices</a>
                 @endcan    
                     <a class="dropdown-item" href="{{ route('invoice.tax-report') }}">Monthly tax report</a>
+                    <a class="dropdown-item" href="{{ route('invoice.invoice-details') }}"> Invoice details listing</a>
```
Comment (GauravGusain98) [inline resources/views/layouts/navbar.blade.php:65]:
```suggestion
                    <a class="dropdown-item" href="{{ route('invoice.details') }}">Invoice details listing</a>
```

File: Modules/Invoice/Resources/views/invoice-details-listing.blade.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+                        {{ $loop->iteration }} 
+                        <td>{{ $invoice->sent_on->toDateString() }}</td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
+                        <td></td>
```
Comment (GauravGusain98) [inline Modules/Invoice/Resources/views/invoice-details-listing.blade.php:48]: Why these columns empty?

Comment (GauravGusain98) [reply Modules/Invoice/Resources/views/invoice-details-listing.blade.php:48]: Please confirm if we need to fill these details or not?

**Review by GauravGusain98 — CHANGES_REQUESTED** [review]
Please check the comments.

File: Modules/Invoice/Exports/MonthlyReportExport.php [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+namespace Modules\Invoice\Exports;
+
+use Maatwebsite\Excel\Concerns\WithTitle;
+use Maatwebsite\Excel\Concerns\WithHeadings;
+use Maatwebsite\Excel\Concerns\FromCollection;
+use Maatwebsite\Excel\Concerns\ShouldAutoSize;
+
+class MonthlyReportExport implements FromCollection, WithHeadings, ShouldAutoSize, WithTitle
```
Comment (P4NK4J) [inline Modules/Invoice/Exports/MonthlyReportExport.php:10]: Can we rename it to something more specific, let's say `MonthlyGSTTaxReportExport`. Let me know what you think

File: Modules/Invoice/Exports/MonthlyReportExport.php [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        if (is_array($this->invoices->first())) {
+            return array_keys($this->invoices->first());
+        }
+
+        return [];
+    }
+
+    public function title(): string
+    {
+        return 'Monthly invoice report';
```
Comment (P4NK4J) [inline Modules/Invoice/Exports/MonthlyReportExport.php:38]: Let's rename it here as well, `Monthly GST Tax Report`

File: Modules/Invoice/Resources/views/invoice-details-listing.blade.php [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+                        <td>{{ $loop->iteration }}</td>
+						<td>{{ $invoice->sent_on->format(config('invoice.default-date-format')) }}</td>
+						<td>{{ $clients[$key]->name }}</td>
+						<td>{{ ($clientAddress[$key]->country_id == 1 ) ? 'India' : 'Export for international invoice'}}</td>
+						<td>{{ $invoice->invoice_number }}</td>
+						<td>{{ ($clientAddress[$key]->country_id == 1 ) ? !empty($invoice->gst) ? $invoice->gst : 'B2C' : 'Export for international invoice' }}</td>
+						<td>{{ $invoice->invoiceAmount() }}</td>
+						<td>{{ $currentRates }}</td>
+						<td>{{ ($clientAddress[$key]->country_id == 2 ) ? $totalReceivableAmount : $invoice->invoiceAmount() }}</td>
+						<td>{{ $invoice->amount }}</td>
```
Comment (P4NK4J) [inline Modules/Invoice/Resources/views/invoice-details-listing.blade.php:50]: The taxable amount should be invoice amount excluding GST for Indian clients, please refer this issue #1340

File: Modules/Invoice/Resources/views/invoice-details-listing.blade.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+						<td>{{ ($invoice->currency == "USD") ? $igst[$key] : '' }}</td>
+						<td>{{ ($invoice->currency == "INR") ? $cgst[$key] : '' }}</td>
+						<td>{{ ($invoice->currency == "INR") ? $sgst[$key] : '' }}</td>
```
Comment (P4NK4J) [inline Modules/Invoice/Resources/views/invoice-details-listing.blade.php:53]: IGST will be implemented when billing state is not Haryana for Indian clients, why USD is checked here ?
Also, CGST and SGST will be added if billing state is Haryana. Please make changes

File: Modules/Invoice/Routes/web.php [hunk 9]
```diff
@@ -23,4 +23,6 @@
     Route::post('/{invoiceId}/update', 'InvoiceController@update')->name('invoice.update');
     Route::delete('/{invoiceId}/delete', 'InvoiceController@destroy')->name('invoice.delete');
     Route::post('/{invoice}', 'InvoiceController@sendEmail')->name('invoice.sendEmail');
+    Route::get('/invoice-details-listing', 'InvoiceController@invoiceDetails')->name('invoice.details');
+    Route::get('/monthly-report-export', 'InvoiceController@monthlyReportExport')->name('invoice.monthly-report-export');
```
Comment (P4NK4J) [inline Modules/Invoice/Routes/web.php:27]: Please provide specific names to the routes and functions here

File: Modules/Invoice/Services/InvoiceService.php [hunk 10]
```diff
@@ -139,6 +141,11 @@ public function getClientsForInvoice()
         return app(ClientServiceContract::class)->getAll();
     }
 
+    public function getClientAddressesForInvoice()
+    {
+        return app(ClientServiceContract::class)->getAll();
```
Comment (P4NK4J) [inline Modules/Invoice/Services/InvoiceService.php:146]: What's the use of this ?

File: Modules/Invoice/Services/InvoiceService.php [hunk 11]
```diff
@@ -213,6 +220,74 @@ public function taxReportExport($filters)
...
+            'clientAddress' => $clientAddress,
+            'currentRates' => $this->currencyService()->getCurrentRatesInINR(),
+            'totalReceivableAmount' => $totalReceivableAmount,
+            'igst' => $igst,
+            'cgst' => $cgst,
+            'sgst' => $sgst
+        ];
+    }
+
+    public function monthlyReportExport($filters)
```
Comment (P4NK4J) [inline Modules/Invoice/Services/InvoiceService.php:253]: The function name gives a very generic sense of monthly report, can we make it more specific in terms of GST tax report ?

File: Modules/Invoice/Services/InvoiceService.php [hunk 12]
```diff
@@ -213,6 +220,74 @@ public function taxReportExport($filters)
...
+    {
+        $invoices = $this->monthlyReportInvoices($filters);
+        $invoices = $this->formatMonthlyInvoicesForExportAll($invoices);
```
Comment (P4NK4J) [inline Modules/Invoice/Services/InvoiceService.php:256]: Please rename this function to something meaningful. Its less readable with the current name

File: Modules/Invoice/Services/InvoiceService.php [hunk 13]
```diff
@@ -213,6 +220,74 @@ public function taxReportExport($filters)
...
+
+        return Excel::download(new MonthlyReportExport($invoices), 'MonthlyReportExport.xlsx');
```
Comment (P4NK4J) [inline Modules/Invoice/Services/InvoiceService.php:258]: Please rename the filename as well

**Review by P4NK4J — CHANGES_REQUESTED** [review]
Left some comments, please check.

**Review by P4NK4J — APPROVED** [review]

**Review by GauravGusain98 — APPROVED** [review]
Looks fine

**Pull request merged**